client/codegen_pinvoke/Main.cs

```diff
@@ -88,24 +89,37 @@ static int Main(string[] args)
                     lib.output_dir.Delete(true);
                 }
 
-                const string single_lib_name = "unity_states";
-
                 // Actually generate the code
-                CppSharp.ConsoleDriver.Run(new Codegen(lib, single_lib_name));
+                var override_lib_name = System.Environment.GetEnvironmentVariable("CONSTELLATION_DLLIMPORT_NAME")!;
+                CppSharp.ConsoleDriver.Run(new Codegen(lib, override_lib_name));
             }
 
             return 0;
         }
 
-        public Codegen(LibInfo lib_info, string single_lib_name)
+        public Codegen(LibInfo lib_info, string override_lib_name)
         {
             this.lib_info = lib_info;
-            this.single_lib_name = single_lib_name;
+            this.override_lib_name = override_lib_name;
         }
 
         /// Setup the driver options here.
         public void Setup(CppSharp.Driver driver)
         {
+            // Copy all DLLs to override location
+            var directoryInfo = new DirectoryInfo(this.lib_info.cargo_artifact_dir.FullName);
+            var filesList = directoryInfo.GetFiles($"lib{this.lib_info.crate_name}.*");

```

Comment:
I am not sure why this file, which should *purely* be in charge of generating code, is doing any copying from other locations. 

I thought we would instead have the user to provide a path to the dylib we are generating code from, and an optional string to change the contents of the DLLImport?

Reply:
That is originally how I was going to implement it, however CppSharp generates the `DllImport` string based on the file name. Other than `GenerateInternalImports` which results in `__Internal`, I didn't see a way to override that string.

See from CppSharp source: [this line](https://github.com/mono/CppSharp/blob/main/src/Generator/Generators/CSharp/CSharpSources.cs#L3497) through the end of the function.

Reply:
> In that case, let's just pass the path of the native lib to the code
generator. The basename of the dylib will determine the name of the
DLLImport.

Yes, we are already passing the path of the native lib to the code generator. The basename does determine the argument to `DllImport`, and that's a problem because the goal is to override the `DllImport` argument. If the filename is `libunity_states.so`, then the DllImport will be `[DllImport("unity_states")]` regardless of what else we pass.

Am I missing a parameter that we can set?

Reply:
> Yes, we are already passing the path of the native lib to the code generator.

What I mean is that currently the code generator script is copying/mutating the input files, moving files, etc. This is confusing as a code generator shouldn't do this - it should be passed an input path, and then generate code from it, nothing more.

If we cannot directly tell the code generator to rename the DLLImport to a string, then it is sufficient to have the name come directly from the name of the native library path. If the user wants a different name, they should rename the native library.

In our case, if we produce `libtp_client.so`, we should copy this file in a step in the movieoke contract csproj, which allows us to rename the file and pass its path to the code generator script.

Reply:
Pushed comments for this based on what we discussed verbally earlier today - keeping the same strategy.